### PR TITLE
vd_lavc: repeatedly attempt to fallback if hwdec fails in receive_frame

### LINKED ISSUE #11947: --hwdec=auto fails to use software decoding
### Important Information

- mpv version: v0.35.0-542-gcab544889a (built via ports) + ffmpeg n6.1-dev-1559-g9557bf26b3
- OS version: FreeBSD 14.0-CURRENT amd64 (built on 2023-07-12)
- WM version: sway 1.8.1.169-g6c234d013 + wlroots 0.16.0-651-g88942d43fd
- GPU version: drm-515-kmod-5.15.25_3 (i915) + mesa-devel-23.2.0.b.106_1 (anv)
- Regressed by 4501e079968d

### Reproduction steps

Requires a GPU that supports a codec but only 8-bit profiles. For example, H.264/AVC High10 on many GPUs or H.265/HEVC Main10 on Intel Skylake.

```
$ mpv --no-config --msg-level=vd=debug,vo=v --hwdec=auto --vo=gpu-next --gpu-api=vulkan \
  http://www.larmoire.info/jellyfish/media/jellyfish-3-mbps-hd-hevc-10bit.mkv
 (+) Video --vid=1 (*) (hevc 1920x1080 29.970fps)
[vo/gpu-next] Probing for best GPU context.
[vo/gpu-next/vulkan] Initializing GPU context 'waylandvk'
[vo/gpu-next/libplacebo] Initialized libplacebo v6.292.0 (API v292)
[vo/gpu-next/libplacebo] Creating vulkan instance with extensions:
[vo/gpu-next/libplacebo]     VK_KHR_get_physical_device_properties2
[vo/gpu-next/libplacebo]     VK_KHR_surface
[vo/gpu-next/libplacebo]     VK_EXT_swapchain_colorspace
[vo/gpu-next/libplacebo]     VK_KHR_external_memory_capabilities
[vo/gpu-next/libplacebo]     VK_KHR_external_semaphore_capabilities
[vo/gpu-next/libplacebo]     VK_KHR_get_surface_capabilities2
[vo/gpu-next/libplacebo]     VK_KHR_portability_enumeration
[vo/gpu-next/libplacebo]     VK_KHR_surface
[vo/gpu-next/libplacebo]     VK_KHR_wayland_surface
[vo/gpu-next/wayland] Registered for protocol wl_shm
[vo/gpu-next/wayland] Registered for protocol zwp_linux_dmabuf_v1
[vo/gpu-next/wayland] Registered for protocol wl_compositor
[vo/gpu-next/wayland] Registered for protocol wl_subcompositor
[vo/gpu-next/wayland] Registered for protocol wl_data_device_manager
[vo/gpu-next/wayland] Registered for protocol zwp_idle_inhibit_manager_v1
[vo/gpu-next/wayland] Registered for protocol xdg_wm_base
[vo/gpu-next/wayland] Registered for protocol zxdg_decoration_manager_v1
[vo/gpu-next/wayland] Registered for protocol wp_presentation
[vo/gpu-next/wayland] Registered for protocol wp_viewporter
[vo/gpu-next/wayland] Registered for protocol wp_single_pixel_buffer_manager_v1
[vo/gpu-next/wayland] Registered for protocol wp_content_type_manager_v1
[vo/gpu-next/wayland] Registered for protocol wp_fractional_scale_manager_v1
[vo/gpu-next/wayland] Registered for protocol wl_seat
[vo/gpu-next/wayland] Registered for protocol wl_output
[vo/gpu-next/wayland] Registered output LG Electronics LG Ultra HD (0x2f):
[vo/gpu-next/wayland]   x: 0px, y: 0px
[vo/gpu-next/wayland]   w: 3840px (600mm), h: 2160px (340mm)
[vo/gpu-next/wayland]   scale: 1
[vo/gpu-next/wayland]   Hz: 60.000000
[vo/gpu-next/wayland] Enabling server decorations
[vo/gpu-next/libplacebo] Probing for vulkan devices:
[vo/gpu-next/libplacebo]     GPU 0: Intel(R) HD Graphics 530 (SKL GT2) v1.3.255 (integrated)
[vo/gpu-next/libplacebo]            uuid: 86:80:12:19:06:00:00:00:00:02:00:00:00:00:00:00
[vo/gpu-next/libplacebo] Vulkan device properties:
[vo/gpu-next/libplacebo]     Device Name: Intel(R) HD Graphics 530 (SKL GT2)
[vo/gpu-next/libplacebo]     Device ID: 8086:1912
[vo/gpu-next/libplacebo]     Device UUID: 86:80:12:19:06:00:00:00:00:02:00:00:00:00:00:00
[vo/gpu-next/libplacebo]     Driver version: 5c02063
[vo/gpu-next/libplacebo]     API version: 1.3.255
[vo/gpu-next/libplacebo] Creating vulkan device with extensions:
[vo/gpu-next/libplacebo]     VK_KHR_swapchain
[vo/gpu-next/libplacebo]     VK_KHR_swapchain
[vo/gpu-next/libplacebo]     VK_KHR_push_descriptor
[vo/gpu-next/libplacebo]     VK_KHR_external_memory_fd
[vo/gpu-next/libplacebo]     VK_EXT_external_memory_dma_buf
[vo/gpu-next/libplacebo]     VK_EXT_external_memory_host
[vo/gpu-next/libplacebo]     VK_KHR_external_semaphore_fd
[vo/gpu-next/libplacebo]     VK_EXT_pci_bus_info
[vo/gpu-next/libplacebo]     VK_EXT_image_drm_format_modifier
[vo/gpu-next/libplacebo]     VK_EXT_shader_atomic_float
[vo/gpu-next/libplacebo]     VK_KHR_video_decode_queue
[vo/gpu-next/libplacebo]     VK_KHR_video_decode_h264
[vo/gpu-next/libplacebo]     VK_KHR_video_decode_h265
[vo/gpu-next/libplacebo]     VK_KHR_video_queue
[vo/gpu-next/libplacebo] Memory heaps supported by device:
[vo/gpu-next/libplacebo]     0: flags 0x1 size   48G
[vo/gpu-next/libplacebo] Memory summary:     0 used     0 res     0 alloc, efficiency 100.00%, utilization 100.00%, max page: 3072M
[vo/gpu-next/libplacebo] shaderc SPIR-V version 1.6 rev 1
[vo/gpu-next/libplacebo] Initialized SPIR-V compiler 'shaderc'
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:765: FINISHME: support more multi-planar formats with DRM modifiers
[vo/gpu-next/libplacebo] GPU information:
[vo/gpu-next/libplacebo]     GLSL version: 450 (vulkan)
[vo/gpu-next/libplacebo]       max_shmem_size:            65536
[vo/gpu-next/libplacebo]       max_group_threads:         1024
[vo/gpu-next/libplacebo]       max_group_size[0]:         1024
[vo/gpu-next/libplacebo]       max_group_size[1]:         1024
[vo/gpu-next/libplacebo]       max_group_size[2]:         1024
[vo/gpu-next/libplacebo]       subgroup_size:             32
[vo/gpu-next/libplacebo]       min_gather_offset:         -32
[vo/gpu-next/libplacebo]       max_gather_offset:         31
[vo/gpu-next/libplacebo]     Limits:
[vo/gpu-next/libplacebo]       thread_safe:               1
[vo/gpu-next/libplacebo]       callbacks:                 1
[vo/gpu-next/libplacebo]       max_buf_size:              51539607552
[vo/gpu-next/libplacebo]       max_ubo_size:              134217728
[vo/gpu-next/libplacebo]       max_ssbo_size:             4294967295
[vo/gpu-next/libplacebo]       max_vbo_size:              51539607552
[vo/gpu-next/libplacebo]       max_mapped_size:           51539607552
[vo/gpu-next/libplacebo]       max_buffer_texels:         134217728
[vo/gpu-next/libplacebo]       align_host_ptr:            4096
[vo/gpu-next/libplacebo]       host_cached:               1
[vo/gpu-next/libplacebo]       max_tex_1d_dim:            16384
[vo/gpu-next/libplacebo]       max_tex_2d_dim:            16384
[vo/gpu-next/libplacebo]       max_tex_3d_dim:            2048
[vo/gpu-next/libplacebo]       blittable_1d_3d:           1
[vo/gpu-next/libplacebo]       buf_transfer:              1
[vo/gpu-next/libplacebo]       align_tex_xfer_pitch:      128
[vo/gpu-next/libplacebo]       align_tex_xfer_offset:     128
[vo/gpu-next/libplacebo]       max_variable_comps:        0
[vo/gpu-next/libplacebo]       max_constants:             18446744073709551615
[vo/gpu-next/libplacebo]       max_pushc_size:            128
[vo/gpu-next/libplacebo]       align_vertex_stride:       1
[vo/gpu-next/libplacebo]       max_dispatch[0]:           65535
[vo/gpu-next/libplacebo]       max_dispatch[1]:           65535
[vo/gpu-next/libplacebo]       max_dispatch[2]:           65535
[vo/gpu-next/libplacebo]       fragment_queues:           1
[vo/gpu-next/libplacebo]       compute_queues:            1
[vo/gpu-next/libplacebo]     External API interop:
[vo/gpu-next/libplacebo]       UUID: 86:80:12:19:06:00:00:00:00:02:00:00:00:00:00:00
[vo/gpu-next/libplacebo]       PCI: 0000:00:02:0
[vo/gpu-next/libplacebo]       buf export caps: 0x9
[vo/gpu-next/libplacebo]       buf import caps: 0x19
[vo/gpu-next/libplacebo]       tex export caps: 0x9
[vo/gpu-next/libplacebo]       tex import caps: 0x9
[vo/gpu-next/libplacebo]       sync export caps: 0x1
[vo/gpu-next/libplacebo]       sync import caps: 0x0
[vo/gpu-next/libplacebo] Available surface configurations:
[vo/gpu-next/libplacebo]     0: VK_FORMAT_R16G16B16A16_SFLOAT            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     1: VK_FORMAT_R16G16B16A16_UNORM             VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     2: VK_FORMAT_A2R10G10B10_UNORM_PACK32       VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     3: VK_FORMAT_B8G8R8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     4: VK_FORMAT_B8G8R8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     5: VK_FORMAT_R8G8B8A8_SRGB                  VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     6: VK_FORMAT_R8G8B8A8_UNORM                 VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo]     7: VK_FORMAT_R5G6B5_UNORM_PACK16            VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vo/gpu-next/libplacebo] Picked surface configuration 1: VK_FORMAT_R16G16B16A16_UNORM + VK_COLOR_SPACE_SRGB_NONLINEAR_KHR
[vd] Container reported FPS: 29.970000
[vd] Codec list:
[vd]     hevc - HEVC (High Efficiency Video Coding)
[vd]     hevc_v4l2m2m (hevc) - V4L2 mem2mem HEVC decoder wrapper
[vd] Opening decoder hevc
[vd] Looking at hwdec hevc-vaapi...
[vo/gpu-next] Loading hwdec drivers for format: 'vaapi'
[vo/gpu-next] Loading hwdec driver 'vaapi'
[vo/gpu-next/vaapi] using libplacebo dmabuf interop
[vo/gpu-next/vaapi] Trying to open a x11 VA display...
[vo/gpu-next/vaapi] Trying to open a wayland VA display...
[vo/gpu-next/vaapi/vaapi] Initialized VAAPI: version 1.19
[vo/gpu-next/vaapi] Going to probe surface formats (may log bogus errors)...
[vo/gpu-next/vaapi] Done probing surface formats.
[vd] Trying hardware decoding via hevc-vaapi.
[vd] Selected codec: hevc (HEVC (High Efficiency Video Coding))
[vd] Pixel formats supported by decoder: vaapi vulkan yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[ffmpeg/video] hevc: No support for codec hevc profile 2.
[vd] Hardware decoding of this stream is unsupported?
[vd] Requesting pixfmt 'yuv420p10le' from decoder.
[vd] Attempting next decoding method after failure of hevc-vaapi.
[vd] Skipping previously attempted hwdec: hevc-vaapi
[vd] Looking at hwdec hevc-vaapi-copy...
[vd] Trying hardware decoding via hevc-vaapi-copy.
[vd] Pixel formats supported by decoder: vaapi vulkan yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[vd] Requesting pixfmt 'vaapi' from decoder.
[ffmpeg/video] hevc: No support for codec hevc profile 2.
[ffmpeg/video] hevc: Failed setup for format vaapi: hwaccel initialisation returned error.
[vd] Pixel formats supported by decoder: vulkan yuv420p10le yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[vd] Requesting pixfmt 'yuv420p10le' from decoder.
[vd] Attempting next decoding method after failure of hevc-vaapi-copy.
[vd] Skipping previously attempted hwdec: hevc-vaapi
[vd] Skipping previously attempted hwdec: hevc-vaapi-copy
[vd] Looking at hwdec hevc-vulkan...
[vo/gpu-next] Loading hwdec drivers for format: 'vulkan'
[vo/gpu-next] Loading hwdec driver 'vulkan'
[vd] Trying hardware decoding via hevc-vulkan.
[vd] Pixel formats supported by decoder: vaapi vulkan yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[vd] Hardware decoding of this stream is unsupported?
[vd] Requesting pixfmt 'yuv420p10le' from decoder.
[vd] Attempting next decoding method after failure of hevc-vulkan.
[vd] Skipping previously attempted hwdec: hevc-vaapi
[vd] Skipping previously attempted hwdec: hevc-vaapi-copy
[vd] Skipping previously attempted hwdec: hevc-vulkan
[vd] Looking at hwdec hevc-vulkan-copy...
[vd] Trying hardware decoding via hevc-vulkan-copy.
[vd] Pixel formats supported by decoder: vaapi vulkan yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[vd] Requesting pixfmt 'vulkan' from decoder.
[ffmpeg/video] hevc: Failed setup for format vulkan: hwaccel initialisation returned error.
[vd] Pixel formats supported by decoder: vaapi yuv420p10le yuv420p10le
[vd] Codec profile: Main 10 (0x2)
[vd] Requesting pixfmt 'yuv420p10le' from decoder.
[vd] Attempting next decoding method after failure of hevc-vulkan-copy.
[vd] Skipping previously attempted hwdec: hevc-vaapi
[vd] Skipping previously attempted hwdec: hevc-vaapi-copy
[vd] Skipping previously attempted hwdec: hevc-vulkan
[vd] Skipping previously attempted hwdec: hevc-vulkan-copy
[vd] Looking at hwdec hevc_v4l2m2m-v4l2m2m-copy...
[vd] Trying hardware decoding via hevc_v4l2m2m-v4l2m2m-copy.
[vd] Using underlying hw-decoder 'hevc_v4l2m2m'
[ffmpeg/video] hevc_v4l2m2m: Could not find a valid device
[ffmpeg/video] hevc_v4l2m2m: can't configure decoder
Could not open codec.
V: 00:00:00 / 00:00:30 (0%) Cache: 0.4s/107KB
[vd] Uninit decoder.

Exiting... (Errors when loading file)
[vo/gpu-next/wayland] Deregistering output LG Electronics LG Ultra HD (0x2f)
```

### Expected behavior
`--hwdec=auto` falls back to software decoding.

### Actual behavior
`--hwdec=auto` tries `vaapi`, `vulkan` then gives up.

### PR DESCRIPTION
**Pull request by philipl**
There is an additional failure path I didn't account for in my previous work. While I ensured that a late hwdec failure in receive_frame can be recovered from my trying the next hwdec, there is a specific combination where if an hwdec fails in receive_frame, and the next hwdec is a full decoder (eg: v4l2m2m), and that also fails, we are left with no decoder and so the entire decoding process ends and playback is stopped.

Basically, we must keep re-attempting the fallback in receive_frame until we get a valid decoder (software or hardware). This edge case will rarely be encountered as there are only a couple of decoder based hwdecs.

Fixes #11947

### CODE REVIEW

**Review by Dudemanguy — APPROVED** [review]
works for me™
